---
title: "Getting Started"
format: 
  html:
    code-fold: false
    code-tools: true
editor: visual
author: "Jenny L. Smith"
---

::: callout-tip
## General Conventions

A description of:

1.  Naming conventions for Github repositories
2.  Using a data analysis notebook template
3.  Initializing the code repo using `git` and `github`
:::

<!-- You can experiment with `code-fold` and `code-tools` in the yaml header above to change how the code cells look like. -->

## Naming Conventions

Repositories have the following nomenclature for easy searching:

1.  Exploratory Data Analysis and Research Projects with collaborators:

-   `[YEAR-MONTH-DAY]_[DATA-TYPE]_[SHORT-DESCRIPTION]`

-   Example: '2025.06.02_RNAseq_Bulk_T-ALL'

-   Example with two data types: '2025.06.02_RNAseq_WGS_Bulk_T-ALL'

2.  Genomic Data pipelines and workflows:

-   `[DATA-TYPE]_[DATA-PROCESSING]_[WORKFLOW-TYPE]`

-   Example: 'RNAseq_Fusion_Calling_Nextflow'

R packages:

-   `[SHORT-NAME]_R_Package`

-   Example: 'DeGSEA_R_Package'

Python packages:

-   `[SHORT-NAME]_Py_Package`

-   Example: 'DataValidator_Py_Package'

## Data Analysis Notebook Template

::: {.panel-tabset}

###  R with rmarkdown and quarto

``` bash        
DATATYPE="RNAseq" 
DESCRIPTION="T-ALL_fusions" 
OUTDIR="$(date +%Y-%m-%d)_${DATATYPE}_${DESCRIPTION}"

mkdir $OUTDIR
cd $OUTDIR
quarto use template Meshinchi-Lab/analysis_template
```

> You will see the following output: \> Downloading \[###################################\] 0.0 \> \[✓\] Downloading \> \[✓\] Unzipping \> Preparing template files... \> \[✓\] Copying files... \> Files created: \> - 2025-06-13_RNAseq_T-ALL_fusions.qmd

The `2025-06-13_RNAseq_T-ALL_fusions.qmd` will be the EDA report. You can rename it if you want. You can now open the `.qmd` file in your favorite text editor and start editing it. When you are ready to render the HTML report, run:

``` bash
quarto render 2025-06-13_RNAseq_T-ALL_fusions.qmd
```

### Python with jupyter notebooks and quarto

Another quarto template!

:::







## Initializing the Repository

basic git command line to commit and push code to the lab github organization

::: callout-note
## Task 1

*This is a callout-note, and it can be quite useful for exercises. You can find more about callout [here](https://quarto.org/docs/authoring/callouts.html).*

Example: Extend the above analysis to cover all genes.
:::


